cmd/metal-api/internal/fsm/fsm.go

```diff
@@ -0,0 +1,303 @@
+	provisioningFSM, event, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	if e.Src != metal.ProvisioningEventMachineReclaim.String() {
+		handleStateTransition(e)
+	} else if event.Time.Sub(*provisioningFSM.container.LastEventTime) > timeOutAfterMachineReclaim {
+		provisioningFSM.container.LastEventTime = &event.Time
+		provisioningFSM.container.FailedMachineReclaim = true
+	}
+
+	provisioningFSM.container.CrashLoop = false
+}
+
+func parseCallbackArguments(args []interface{}) (*ProvisioningFSM, *metal.ProvisioningEvent, error) {
+	if len(args) != 2 {
+		return nil, nil, errors.New("expecting two arguments of types *ProvisioningFSM, *metal.ProvisioningEvent")
+	}
+
+	provisioningFSM, ok := args[0].(*ProvisioningFSM)
+	if !ok {
+		return nil, nil, errors.New("first argument must be of type *ProvisioningFSM")
+	}
+
+	event, ok := args[1].(*metal.ProvisioningEvent)
+	if !ok {
+		return nil, nil, errors.New("second argument must be of type *metal.ProvisioningEvent")
+	}
+
+	return provisioningFSM, event, nil
+}
+
+func newProvisioningFSM(initialState metal.ProvisioningEventType, container *metal.ProvisioningEventContainer) *ProvisioningFSM {
+	f := fsm.NewFSM(
+		initialState.String(),
+		events,
+		callbacks,
+	)
+
+	provisioningFSM := ProvisioningFSM{
+		fsm:       f,
+		container: container,
+	}
+
+	return &provisioningFSM
+}
+
+// HandleProvisioningEvent writes the ProvisioningEvent to the the ProvisioningEventContainer and checks if it is a valid event in the current state
+func HandleProvisioningEvent(event *metal.ProvisioningEvent, container *metal.ProvisioningEventContainer, log *zap.SugaredLogger) (*metal.ProvisioningEventContainer, error) {
+	if len(container.Events) == 0 {
+		container.Events = append(container.Events, *event)
+		container.LastEventTime = &(*event).Time
+		container.Liveliness = metal.MachineLivelinessAlive
+
+		return container, nil
+	}
+
+	provisioningFSM := newProvisioningFSM(container.Events[len(container.Events)-1].Event, container)
+
+	err := provisioningFSM.fsm.Event(event.Event.String(), provisioningFSM, event)
+	if err != nil {
+		if errors.Is(err, fsm.NoTransitionError{}) {
+			err = nil
+			container.LastEventTime = &(*event).Time
+			container.Liveliness = metal.MachineLivelinessAlive
+			if event.Event == metal.ProvisioningEventAlive {
+				log.Debugw("received provisioning alive event", "id", container.ID)
+
+			} else if event.Event == metal.ProvisioningEventPhonedHome {
+				log.Debugw("swallowing repeated phone home event", "id", container.ID)
+			}
+		} else if errors.As(err, &fsm.InvalidEventError{}) {
+			container.Events = append(container.Events, *event)
+			container.LastEventTime = &(*event).Time
+			container.CrashLoop = true
+			err = nil
+		}
+	}
+
+	return container, err
+}
+
+func ProvisioningEventForMachine(log *zap.SugaredLogger, ec *metal.ProvisioningEventContainer, machineID, event, message string) *metal.ProvisioningEventContainer {
```

Comment:
I think it's better to pass the `metal.ProvisioningEvent` and not event and message strings. One less func arg and caller has the required information.

Also you probably don't need to pass the machineID as it should be contained in the provisioning event container. Let the caller pass it properly and don't allow it to be nil by returning an error.